### PR TITLE
update mode if group or owner change to keep suid bit

### PR DESCRIPTION
**Pull request by minshallj**
On Linux updating the group or owner unsets the suid bit for security
reasons, so check for group and owner updates whether or not to set the
mode.

This is in reference to issue #2951

### CODE REVIEW

**Comment by minshallj** [comment]
Is the appveyor failure my fault? I would doubt it because it fails very early, it it's a failure to install a gem. Plus it looks like it's on windows and all of my changes were under unix only areas.

**Comment by nathenharvey** [comment]
The AppVeyor failure looks like an actual failure with the file_access_control_spec.rb.  Can you please look into that and update your code or the tests appropriately?

**Comment by minshallj** [comment]
I can try looking into this, but I don't have a windows machine to test on, and when I `bundle exec rspec ./spec/unit/file_access_control_spec.rb` all tests pass. I do see the error in the test now though, so when I have time I'll try pushing some commits to fix it.

**Comment by minshallj** [comment]
So looking around at this failure I noticed
1. There was a problem with my boolean logic, which I am pushing a patch for now
2. The resource being created didn't have a user or group. I am wondering if this is only something that might happen in a test case since we are really just testing to see if the mode _should_ be changed here.

Pushing the commit now, and it might fix this break because of the boolean logic (once it notices the suid bit is not set it will no longer check the conditions on the far right) The root cause of this failure though is that a user/group should be set when checking if a resource _should_ have it's mode changed now.

**Comment by minshallj** [comment]
Ping.

I guess some questions that might still remain about this patch are, "Is this the best way to check for the suid bit?", "This isn't needed on macs, can we avoid doing this there", and lastly "can we suppress the output of 'changed mode from 04755 to 04755' since it seems redundant".

1. Well, the mode as a string will always have the owner, group, and global permissions as the last three characters of the string respectively. Thus if the suid bit is set, the fourth character from the right would be something other than 0. I'm not totally sure if the mode as a string will always be 4+ characters long, if it has the possibility of being \< 3 characters, the suid check might want to look like `return mode_to_s(target_mode) and mode_to_s(target_mode)[-4] != '0'`. I can definitely add that in if requested.
2. This shouldn't be hard, though I'm not totally sure this isn't needed on macs, and that it won't be in the future. Something like `!(/darwin/ =~ RUBY_PLATFORM)` could be thrown in the if statement, again easy to accomplish if desired.
3. No idea how to go about this. Is there some way to suppress the output?

File: lib/chef/file_access_control/unix.rb [hunk 1]
```diff
@@ -280,6 +282,9 @@ def uid_from_resource(resource)
         return nil
       end
 
+      def suid_bit_set
+          return mode_to_s(target_mode)[-4] != '0'
```
Comment (bahamas10) [inline lib/chef/file_access_control/unix.rb]: this means the setuid, setgid, or the sticky-bit (or a combination of them) is set... if you are *only* concerned about the setuid bit you should check for explicitly (also without casting to a string)

``` ruby
def suid_bit_set
  target_mode & 04000 > 0
end
```

**Comment by minshallj** [comment]
@bahamas10 ok, made that change and squashed the non test related commits into one

**Comment by minshallj** [comment]
@nathenharvey, any opposition to merging this?

**Comment by thommay** [comment]
:+1: @chef/client-core

**Comment by danielsdeleo** [comment]
:+1: as long as we fix up the indentation on merge.

**Pull request merged**